File: src/main/java/com/braydenoneal/blang/parser/expression/builtin/list/ListRemoveBuiltin.java
package com.braydenoneal.blang.parser.expression.builtin.list;

import com.braydenoneal.blang.parser.Program;
import com.braydenoneal.blang.parser.expression.Expression;
import com.braydenoneal.blang.parser.expression.value.IntegerValue;
import com.braydenoneal.blang.parser.expression.value.ListValue;
import com.braydenoneal.blang.parser.expression.value.Value;

import java.util.List;

public record ListRemoveBuiltin(
        String name,
        ListValue listValue,
        List<Expression> arguments
) implements Expression {
    @Override
    public Value<?> evaluate(Program program) {
        Value<?> removeValue = arguments.getFirst().evaluate(program);

        List<Value<?>> localList = listValue.value();

        if (removeValue instanceof IntegerValue integerValue) {
            localList.remove((int) integerValue.value());
        } else {
            localList.remove(removeValue);
        }

        return program.getScope().set(name, new ListValue(localList));
    }
}

File: src/main/kotlin/com/braydenoneal/blang/parser/expression/builtin/list/ListRemoveBuiltin.kt
package com.braydenoneal.blang.parser.expression.builtin.list

import com.braydenoneal.blang.parser.Program
import com.braydenoneal.blang.parser.expression.Arguments
import com.braydenoneal.blang.parser.expression.Expression
import com.braydenoneal.blang.parser.expression.ExpressionType
import com.braydenoneal.blang.parser.expression.ExpressionTypes
import com.braydenoneal.blang.parser.expression.value.IntegerValue
import com.braydenoneal.blang.parser.expression.value.ListValue
import com.braydenoneal.blang.parser.expression.value.Value
import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder


data class ListRemoveBuiltin(
    val listValue: ListValue,
    val arguments: Arguments
) : Expression {
    override fun evaluate(program: Program): Value<*> {
        val removeValue = arguments.anyValue(program, "value", 0)

        if (removeValue is IntegerValue) {
            listValue.value.removeAt(removeValue.value)
        } else {
            listValue.value.remove(removeValue)
        }

        return listValue
    }

    override val type: ExpressionType<*> get() = ExpressionTypes.LIST_REMOVE_BUILTIN

    companion object {
        val CODEC: MapCodec<ListRemoveBuiltin> = RecordCodecBuilder.mapCodec {
            it.group(
                ListValue.CODEC.fieldOf("listValue").forGetter(ListRemoveBuiltin::listValue),
                Arguments.CODEC.fieldOf("arguments").forGetter(ListRemoveBuiltin::arguments)
            ).apply(it, ::ListRemoveBuiltin)
        }
    }
}
